Property type constraints now accepts lists as well (#646)
neo4j/docs-cypher: modules/ROOT/pages/constraints/examples.adoc

[[constraints-examples-node-property-type]]
== Node property type constraints

> _This feature was introduced in Neo4j 5.9 and extended in 5.10._

Reviewer: ```suggestion
_This feature was introduced in Neo4j 5.9._
```

* `ZONED DATETIME`
* `DURATION`
* `POINT`

> * `LIST<BOOLEAN NOT NULL>`

Reviewer 2: Do we need to flag these as introduced in 5.10 somehow? The above implies that they are available in 5.9.

======

[discrete]
[[constraints-create-node-property-type-constraints-on-invalid-types]]
=== Creating constraints on invalid types will fail

.+CREATE CONSTRAINT+
======

Create a node property type constraint restricting the property `imdbScore` to map values on nodes with the `Movie` label:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score
FOR (movie:Movie) REQUIRE movie.imdbScore IS :: MAP
----

> In this case, the constraint cannot be created because the `MAP` type is not allowed in property type constraints.

Reviewer 2: ```suggestion
This constraint cannot be created because values of type `MAP` cannot be stored in properties and therefore are not permitted in property type constraints.
```

Author: I took the `In this case, ` part from existing failing examples so it might be nice to keep that for consistency?
```suggestion
In this case, the constraint cannot be created because values of type `MAP` cannot be stored in properties and therefore are not permitted in property type constraints.
```

Reviewer 2: Sure, we can keep it. I removed it because it seems unnecessary.

.Error message
[source, error]
----
Failed to create node property type constraint: Invalid property type `MAP`.
----

======

.+CREATE CONSTRAINT+
======

Create a node property type constraint restricting the property `imdbScore` to list of nullable float values on nodes with the `Movie` label:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score
FOR (movie:Movie) REQUIRE movie.imdbScore IS :: LIST<FLOAT>
----

> In this case, the constraint cannot be created because the inner type of list types cannot be nullable.

Reviewer 2: ```suggestion
In this case, the constraint cannot be created because the inner type of list types cannot be nullable. The correct type to use for the constraint is `LIST<FLOAT NOT NULL>` because null values cannot be stored as part of a list.
```

Author: this has been added as line 1300

.Error message
[source, error]
----
Failed to create node property type constraint: Invalid property type `LIST<FLOAT>`. Lists cannot have nullable inner types.
----

======

.+CREATE CONSTRAINT+
======

Create a node property type constraint restricting the property `imdbScore` to list of lists of float values on nodes with the `Movie` label:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score

> FOR (movie:Movie) REQUIRE movie.imdbScore IS :: LIST<LIST<FLOAT>>

Reviewer 2: Let's make this NOT NULL so as not to confuse with the previous case. 
```suggestion
FOR (movie:Movie) REQUIRE movie.imdbScore IS :: LIST<LIST<FLOAT NOT NULL>>
```

----

In this case, the constraint cannot be created because the inner type of list types cannot be other lists.

.Error message
[source, error]
----

> Failed to create node property type constraint: Invalid property type `LIST<LIST<FLOAT>>`. Lists cannot have lists as an inner type.

Reviewer 2: ```suggestion
Failed to create node property type constraint: Invalid property type `LIST<LIST<FLOAT NOT NULL>>`. Lists cannot have lists as an inner type.
```

[[constraints-examples-relationship-property-type]]
== Relationship property type constraints

> _This feature was introduced in Neo4j 5.9 and extended in 5.10._

Reviewer: ```suggestion
_This feature was introduced in Neo4j 5.9._
```

* `ZONED DATETIME`
* `DURATION`
* `POINT`

> * `LIST<BOOLEAN NOT NULL>`

Reviewer 2: Same as the above comment, should we flag these as added in 5.10 somehow?

=== Creating constraints on invalid types will fail

.+CREATE CONSTRAINT+
======

Create a relationship property type constraint restricting the property `releaseOrder` to integer values excluding `null` on relationships with the `PART_OF` relationship type:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score
FOR ()-[part:PART_OF]-() REQUIRE part.releaseOrder IS :: INTEGER NOT NULL
----

In this case, the constraint cannot be created because excluding `null` is not allowed in property type constraints.
To also ensure that the property exists (is not `null`), add an existence constraint on the property.

.Error message
[source, error]
----
Failed to create relationship property type constraint: Invalid property type `INTEGER NOT NULL`.
----

======

.+CREATE CONSTRAINT+
======

Create a relationship property type constraint restricting the property `releaseOrder` to list of nullable integer values on relationships with the `PART_OF` relationship type:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score
FOR ()-[part:PART_OF]-() REQUIRE part.releaseOrder IS :: LIST<INTEGER>
----

> In this case, the constraint cannot be created because the inner type of list types cannot be nullable.

Reviewer 2: ```suggestion
In this case, the constraint cannot be created because the inner type of list types cannot be nullable. The correct type to use for the constraint is `LIST<INTEGER NOT NULL>` because null values cannot be stored as part of a list.
```

Author: this has been added as line 1633

.Error message
[source, error]
----
Failed to create relationship property type constraint: Invalid property type `LIST<INTEGER>`. Lists cannot have nullable inner types.
----

======

.+CREATE CONSTRAINT+
======

Create a relationship property type constraint restricting the property `releaseOrder` to list of lists of integer values on relationships with the `PART_OF` relationship type:

.Query
[source, cypher, role=test-fail]
----
CREATE CONSTRAINT score

> FOR ()-[part:PART_OF]-() REQUIRE part.releaseOrder IS :: LIST<LIST<INTEGER>>

Reviewer 2: As above, use INTEGER NOT NULL so there is only one thing to fail.
